content/5_Plat/2DRQ.mdx

```diff
@@ -198,20 +199,215 @@ $\mathcal{O}(N^2)$ memory, which is too much.
 Since we know all of the updates and queries beforehand, we can reduce the
 memory usage while maintaining a decent constant factor.
 
-#### Idea 1 - Use an unordered map instead of a 2D array.
+We could use an unordered map instead of a 2D array, but this gives
+$\mathcal{O}(N\log^2N)$ memory and time and the constant factors for both are terrible.
 
-Bad idea ... This gives $\mathcal{O}(N\log^2N)$ memory and time and the constant
-factors for both are terrible.
-
-#### Idea 2 - Compress the points to be updated so that you only need $\mathcal{O}(N\log N)$ memory.
+A better idea is to compress the points to be updated so that you only need $\mathcal{O}(N\log N)$ memory.
 
 This doesn't require knowing the queries beforehand.
 
-- [my 1D offline BIT](<https://github.com/bqi343/USACO/blob/master/Implementations/content/data-structures/1D%20Range%20Queries%20(9.2)/BIToff.h>)
-- [my 2D offline BIT](<https://github.com/bqi343/USACO/blob/master/Implementations/content/data-structures/2D%20Range%20Queries%20(15.2)/BIT2DOff%20(15.2).h>)
+- [Benq's 1D offline BIT](<https://github.com/bqi343/USACO/blob/master/Implementations/content/data-structures/1D%20Range%20Queries%20(9.2)/BIToff.h>)
+- [Benq's 2D offline BIT](<https://github.com/bqi343/USACO/blob/master/Implementations/content/data-structures/2D%20Range%20Queries%20(15.2)/BIT2DOff%20(15.2).h>)
+
+Here's an implementation of the offline 2D BIT presented above that may be
+easier to understand, albeit significantly slower due to a high constant factor:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+// see code example below for usage
+struct BIT2D {
+	int n;
+	vector<vector<int>> vals;
+	vector<vector<int>> bit;
+
+	BIT2D(int n, vector<pair<int, int>> todo) : n(n), vals(n), bit(n) {
```

Comment:
try passing it by reference, maybe that'll speed things up

Reply:
definitely nowhere close to within time limit  but good point lol